rust/lance-index/src/scalar/inverted.rs

```diff
@@ -0,0 +1,513 @@
+
+        for row_ids in &self.row_ids_list {
+            let builder = row_ids_col.values();
+            for row_id in row_ids {
+                builder.append_value(*row_id);
+            }
+            row_ids_col.append(true);
+        }
+
+        for frequencies in &self.frequencies_list {
+            let builder = frequencies_col.values();
+            for frequency in frequencies {
+                builder.append_value(*frequency);
+            }
+            frequencies_col.append(true);
+        }
+
+        let schema = arrow_schema::Schema::new(vec![
+            arrow_schema::Field::new(TOKEN_ID_COL, DataType::UInt32, false),
+            arrow_schema::Field::new(
+                ROW_ID,
+                DataType::List(Field::new_list_field(DataType::UInt64, true).into()),
+                false,
+            ),
+            arrow_schema::Field::new(
+                FREQUENCY_COL,
+                DataType::List(Field::new_list_field(DataType::UInt64, true).into()),
+                false,
+            ),
+        ]);
+
+        let batch = RecordBatch::try_new(
+            Arc::new(schema),
+            vec![
+                Arc::new(token_id_col) as ArrayRef,
+                Arc::new(row_ids_col.finish()) as ArrayRef,
+                Arc::new(frequencies_col.finish()) as ArrayRef,
+            ],
+        )?;
+        Ok(batch)
+    }
+
+    async fn load(reader: Arc<dyn IndexReader>) -> Result<Self> {
+        let mut tokens = Vec::new();
+        let mut row_ids_list = Vec::new();
+        let mut frequencies_list = Vec::new();
+        for i in 0..reader.num_batches().await {
+            let batch = reader.read_record_batch(i).await?;
+            let token_col = batch[TOKEN_ID_COL].as_primitive::<datatypes::UInt32Type>();
+            let row_ids_col = batch[ROW_ID].as_list::<i32>();
+            let frequencies_col = batch[FREQUENCY_COL].as_list::<i32>();
+
+            tokens.extend(token_col.iter().map(|v| v.unwrap()));
+            for value in row_ids_col.iter() {
+                let value = value.unwrap();
+                let row_ids = value
+                    .as_primitive::<datatypes::UInt64Type>()
+                    .values()
+                    .iter()
+                    .cloned()
+                    .collect_vec();
+                row_ids_list.push(row_ids);
+            }
+            for value in frequencies_col.iter() {
+                let value = value.unwrap();
+                let frequencies = value
+                    .as_primitive::<datatypes::UInt64Type>()
+                    .values()
+                    .iter()
+                    .cloned()
+                    .collect_vec();
+                frequencies_list.push(frequencies);
+            }
+        }
+
+        Ok(Self {
+            tokens,
+            row_ids_list,
+            frequencies_list,
+        })
+    }
+
+    fn add(&mut self, token_id: u32, row_id: u64, frequency: u64) {
```

Comment:
I think `frequency` is always 1?  Also, frequency is never used?

Reply:
right, fixed